pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -338,15 +340,111 @@ func (pl *VolumeRestrictions) EventsToRegister() []framework.ClusterEventWithHin
 		// Pods may fail to schedule because of volumes conflicting with other pods on same node.
 		// Once running pods are deleted and volumes have been released, the unschedulable pod will be schedulable.
 		// Due to immutable fields `spec.volumes`, pod update events are ignored.
-		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}},
+		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}, QueueingHintFn: pl.isSchedulableAfterPodDeleted},
 		// A new Node may make a pod schedulable.
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterNodeAdded},
 		// Pods may fail to schedule because the PVC it uses has not yet been created.
 		// This PVC is required to exist to check its access modes.
-		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeClaimChange},
 	}
 }
 
+// isSchedulableAfterPodDeleted is invoked whenever a pod deleted,
+// It checks whether the deleted pod will conflict with volumes of other pods on the same node
+func (pl *VolumeRestrictions) isSchedulableAfterPodDeleted(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	oldPod, _, err := util.As[*v1.Pod](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterPodDeleted", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	if pl.verifyPodPVCConflict(logger, pod) {
+		return framework.QueueAfterBackoff
+	}
+
+	nodeInfo := framework.NewNodeInfo(oldPod)
+	if !satisfyVolumeConflicts(pod, nodeInfo) {
+		return framework.QueueAfterBackoff
+	}
+
+	return framework.QueueSkip
+}
+
+// isSchedulableAfterNodeAdded is invoked whenever a node added, It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeRestrictions) isSchedulableAfterNodeAdded(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	_, newNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterNodeAdded", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	if pl.verifyPodPVCConflict(logger, pod) {
+		return framework.QueueAfterBackoff
```

Comment:
And why? It will fail the next scheduling cycle again because it has a conflict rwop pvc? Why don't we put the pod to backoffQ/activeQ when it's ready?

Reply:
`framework.NewStatus` is not allowed because it is not the same type as `framework.QueueingHint`

Reply:
> `framework.NewStatus` is not allowed because it is not the same type as `framework.QueueingHint`

This is the code in `Filter` indicating that we'll fail in `Filter` if the pod has a conflicting rwop pvc.
Now if this pod is in unschedulable pod pool. And we return `backoff` here, I think we'll put a pod with conflicting rwop pvc to the backOffQ and it will soon fail after backoff. Am I understanding wrong?🤔️

Reply:
The situation with clusters changes from moment to moment, It might become schedulable in the next moment because the pod containing the conflicting pvc is deleted.